benchmark/main.cpp

```diff
@@ -56,979 +50,1198 @@
+
+exit:
+        mbedtls_ecdsa_free(&ecdsa);
     }
-#endif
+#endif /* MBEDTLS_ECDSA_C && MBEDTLS_SHA2565_C */
 
+
+    return ret;
+}
+
+MBED_NOINLINE static int benchmark_ecdh() {
+    int ret = 0;
 #if defined(MBEDTLS_ECDH_C)
-    if( todo->ecdh )
-    {
-        mbedtls_ecdh_context ecdh;
-#if defined(MBEDTLS_ECP_DP_CURVE25519_ENABLED)
-        mbedtls_mpi z;
-#endif
-        const mbedtls_ecp_curve_info *curve_info;
-        size_t olen;
-
-        for( curve_info = mbedtls_ecp_curve_list();
-             curve_info->grp_id != MBEDTLS_ECP_DP_NONE;
-             curve_info++ )
-        {
-            mbedtls_ecdh_init( &ecdh );
-
-            if( mbedtls_ecp_group_load( &ecdh.grp, curve_info->grp_id ) != 0 ||
-                mbedtls_ecdh_make_public( &ecdh, &olen, buf, sizeof( buf ),
-                                  myrand, NULL ) != 0 ||
-                mbedtls_ecp_copy( &ecdh.Qp, &ecdh.Q ) != 0 )
-            {
-                return( 1 );
-            }
-            ecp_clear_precomputed( &ecdh.grp );
-
-            mbedtls_snprintf( title, sizeof( title ), "ECDHE-%s",
-                                              curve_info->name );
-            TIME_PUBLIC( title, "handshake",
-                    ret |= mbedtls_ecdh_make_public( &ecdh, &olen, buf, sizeof( buf ),
-                                             myrand, NULL );
-                    ret |= mbedtls_ecdh_calc_secret( &ecdh, &olen, buf, sizeof( buf ),
-                                             myrand, NULL ) );
-            mbedtls_ecdh_free( &ecdh );
+    mbedtls_ecdh_context ecdh;
+    const mbedtls_ecp_curve_info *curve_info;
+    size_t olen;
+
+    for (curve_info = mbedtls_ecp_curve_list();
+         curve_info->grp_id != MBEDTLS_ECP_DP_NONE && ret == 0;
+         curve_info++) {
+        mbedtls_ecdh_init(&ecdh);
+
+        ret = mbedtls_ecp_group_load(&ecdh.grp, curve_info->grp_id);
+        if (ret != 0) {
+            mbedtls_printf("mbedtls_ecp_group_load() returned -0x%04X\n",
+                           -ret);
+            goto exit;
         }
 
-        /* Curve25519 needs to be handled separately */
-#if defined(MBEDTLS_ECP_DP_CURVE25519_ENABLED)
-        mbedtls_ecdh_init( &ecdh );
-        mbedtls_mpi_init( &z );
+        ret = mbedtls_ecdh_make_public(&ecdh, &olen, buf, sizeof(buf),
+                                       myrand, NULL);
+        if (ret != 0) {
+            mbedtls_printf("mbedtls_ecdh_make_public() returned -0x%04X\n",
+                           -ret);
+            goto exit;
+        }
 
-        if( mbedtls_ecp_group_load( &ecdh.grp, MBEDTLS_ECP_DP_CURVE25519 ) != 0 ||
-            mbedtls_ecdh_gen_public( &ecdh.grp, &ecdh.d, &ecdh.Qp, myrand, NULL ) != 0 )
-        {
-            return( 1 );
+        ret =mbedtls_ecp_copy( &ecdh.Qp, &ecdh.Q );
+        if (ret != 0) {
+            mbedtls_printf("mbedtls_ecp_copy() returned -0x%04X\n", -ret);
+            goto exit;
         }
 
-        TIME_PUBLIC(  "ECDHE-Curve25519"
```

Comment:
I know this is the original behaviour, but if we are alreadey doing a rewrite, please don't exit on failure, we should continue to next test.